### PR TITLE
Fix #1118: follow only dir-symlinks

### LINKED ISSUE #1118: cd: <symlink target file>: Not a directory
File `pyenv` assumes, that it can follow symlinks in `abs_dirname()`:
```bash
  abs_dirname() {
    local path="$1"

    # Use a subshell to avoid changing the current path
    (
    while [ -n "$path" ]; do
      cd "${path%/*}"
      local name="${path##*/}"
      path="$(resolve_link "$name" || true)"
    done

    pwd
    )
  }
```

However, the target file can be a symlink, and it cannot be followed.

Example from Ansible. Listing of `ansible/bin/`:
```
    ansible
    ansible-playbook -> ansible
```

Executing `~/ansible/bin/ansible-playbook` will work, but during it a `~/.pyenv/libexec/pyenv: line 44: cd: ansible: Not a directory` is emitted.

A suggested fix would be:
```bash
      path="$(test -d "$name" && resolve_link "$name" || true)"
```
... to test if the destination is a directory that can be `cd`d into.
If yes, follow that. If not, stop iterating without emitting errors.

### PR DESCRIPTION
**Pull request by webknjaz**
This corrects the deffect where pyenv tries to determine an absolute
path of the directory containing a file and follows symlinks
manually. Because of missing check it behaved correctly, but a
misleading error message was shown in the terminal output.
Resolves #1118.

Thanks to @HQJaTu for the initial patch.

### CODE REVIEW

**Comment by webknjaz** [comment]
@joshfriend @yyuu @timsavage what is the policy for bugfix releases? can we have this out asap plz?

**Comment by webknjaz** [comment]
Ping @joshfriend @yyuu @timsavage

**Comment by timsavage** [comment]
Please stop referencing me, I'm not part of the pyenv team.

**Comment by webknjaz** [comment]
Oops, sorry. I just saw your commits in master and missed they they're coming from PR :)

File: libexec/pyenv [hunk 1]
```diff
@@ -40,7 +40,7 @@ else
 
     # Use a subshell to avoid changing the current path
     (
-    while [ -n "$path" ]; do
+    while [[ -n "$path" && $(echo "$path" | grep /) ]]; do
```
Comment (yyuu) [inline libexec/pyenv]: This will return "pwd" if "$path" is /. At least it must not be the correct return value of the function.

Comment (yyuu) [reply libexec/pyenv]: I think some fix like following will work sufficiently. Please consider adding tests for it.

'''
while [[ "$path" == *"/"* ]]; do
  cd "${path%/*}"
  local name="${path##*/}"
  [ -n "$name" ] || break
  path="$(resolve_link "$name" || true)"
done
'''

Comment (webknjaz) [reply libexec/pyenv]: Thanks! I'll check this

Comment (webknjaz) [reply libexec/pyenv]: Just checked it: it doesn't work your way at all. Will need to investigate some better way.

**Comment by blueyed** [comment]
I've started a while back to fix this in rbenv (upstream), but it needs updating: https://github.com/rbenv/rbenv/pull/868.

**Comment by blueyed** [comment]
btw: `src/configure && make -C src` in pyenv's checkout should fix / work around this, and improve performance as a side-effect.

**Comment by webknjaz** [comment]
Wow, why?

*/me needs to get back to fixing this*

**Pull request closed without merging**